README.md
# shopplinglist
A web based shoppinglist with real time update of the list via Web Sockets.

Screenshot:

![Image seems to be missing][screenshot]

[screenshot]: screenshot.png "Using the theme Sa Barca de Formentera"


logs/changelog.txt
Beskriver projektet "inköpslistan".


* 2016-03-18 18:35
    - Lagt till snyggare favicon, mainfest.json (för proffsigare add to homescreen för Android med ikoner). Theme-color meta-tagg för mobile-index

* 2016-03-06 18:31
    - Lite buggutskrifter som hjälpte mig att få listan att funka på nginx, låter
        de vara kvar.

* 2015-06-21 15:13
	- Använder nu ws://listan.axellarsson.nu istället för ws://axellarsson.nu vilket bör ge mer konsekvent beteende.
	
* 2014-12-27 16:13
    - Att ta bort den sista posten implicerar nu ett anrop av "shoppinglistSocket.clear()" -> detta borde reducera problemet
    med itemIdXX > itemId99 drastiskt.
    
* 2014-12-11 21:20
    - använder nu inte http://list.axellarsson.nu utan /mobile_index.php så att det spelar inte någon roll var listan
    ligger den kommer ändå att fungera för mobiler. Givetvis måste dock ws://axellarsson.nu:9001 fortfarande var hådkodat.

* 2014-11-27 18:26
    - använder nu http://listan.axellarsson.nu och ws://axellarsson.nu:9001

* 2014-08-09 18:28
    - refaktoriserat namnen på teman, ie "Nytt tema" -> Minimalism
    - lagt till ett nytt tema "Sa Barce de Formentera" (baseras på Minimalism)

* 2014-05-31 16:25
    - fixat bugg att updateItem inte fungerade korrekt med isMarked
    - refaktoriserat namn, från gui/oldGui till shoppinglist respektive legacyShoppinglist, även bytt motsvarande filnamn
    
* 2014-05-29 16:11
    - fixade bugg att Safari inte visade Roboto
    - fixade bugg att Firefox på min MacBook visade Roboto Thin Italic istället för
        bara Roboto Thin (den valde automatiskt den installerade Roboto Thin Italic, men  nu har jag också installerat Roboto Thin)
        
* 2014-05-29 15:41
    - fixat att man inte kunde få updateItem att fungera korrekt när man ändrade på mobil, då upptäcktes inte uppdatering av andra mobilklienter
    - fixat 300 ms delay gone away så nu är mobilsidan snabbare än någonsin
    
* 2014-05-29 15:03
    - fixat inloggningen igen iom uppgraderingen av PHP
    - alla mysql-queries är nu uppdaterade till PDO
    - ändrat på localOnly till att endast vara en CSS-klass -> blir mycket lättare
    
* 2014-05-09 18:47
    - syncToDatabase() fungerar nu både upp och ner så att säga...
    
* 2014-04-22 16:38
    - Implementerat removeItem för WebSockets, nu fungerar det mesta bra för Desktop med WebSocket-stöd.
    
* 2014-04-20 XX:XX
    - Fixat animationer igen vid användning av WebSockets.
    
* 2014-04-19 21:48
    - Fixade äcklig bugg i utils.getNextItemId när listan var tom, då blev det ett tyst fel när den inte kunde substringa "undefined".

* 2014-04-19 18:36
    - WebSocket-implementering är klar för Desktop gällande: rensa, lägga till ny, uppdatera.

* 2014-03-24 20:00
	- Påbörjat implementeringen av WebSocket-stöd.
	- Namespace:at javascript.js till oldGUI - skall refaktorisera mer här.
    
* 2013-09-29 20:54
	- Fixat animering när man tar bort och lägger till poster samt när man rensar databasen.
	- Undersökt och researchat Web Sockets vs polling etc.

* 2013-09-28 18:21 
	- Fixat overflow av poster i inköpslistan på Android Stock Browser.

* 2013-09-28 17:57
	- Fixat styling av temaväljaren på mobilenheter.
	
* 2013-09-26 23:56
	- Använder nu jQuery för att flytta "Logga ut"-knappen på mobiler, så som jag gjort med Desktop, detta för att det klassiska temat fortfarande skall fungera.
	
* 2013-09-22 18:50
	- Fixat nytt tema för både login.php och index.php
	
* 2013-09-22 16:09
	- Fixat css-switcher som fungerar konsistent både för mobiler och desktop.
	- Man kan använda ovan nämnda css-switcher för att välja mellan två teman -
		det gamla vanliga klassiska och ett nytt, kallat "Nytt tema"

* 2013-09-13 19:05
	- Projektet använder nu GitHub
	
* 2013-09-08 21:39
	- Fixat den nya designen även för "edit-mode" för Chrome på mobiler.

* 2013-09-07 20:38
	- Implementerat den nya designen av mobilsidan även för "edit-mode". Tog bort alternativet att rensa
		databasen i denna vy - det fick inte plats på ett snyggt sätt och behövs inte för mobilsidan. Än
		så länge fungerar detta endast bara för "Stock Android Browser" ej Chrome för mobiler.

* 2013-09-07 17:05
	- Komplett refaktorisering av hela projektet, allt ligger inte längre huller om buller i rot-mappen - 
		nu är det snyggt organiserat under assets/css, assets/images osv.

* 2013-09-05 20:38
	- Påbörjade en ny design av mobil-sidan, en icke-skeumorfisk sådan. Lite Android-inspirerad med Roboto-font
		etc.

* 2012-08-04 18:42
	- Fixade en bugg som förhindrade att man i vissa, relativt vanliga situationer, inte kunde
		lägga till fler "items". Detta berodde på att listan sorterades på itemId utan tanke på
		inledande nollor. Således sorterades det 1, 11, 12, ..., 2, 21..,8,9 etc. Detta gjorde att
		När man lade till nya items så trodde sidan att det senaste itemId som gällde var item9. Men
		item10 fanns då redan - vilket förhindrade att det kunde läggas till.
	
	- Fixade sorteringsbugg - i och med ovanstående bugg. Som sagt så sorterades tolv varor så här:
		1,11,12,2,3,4,5,6,7,8,9. Nu sorteras de 01,02,03,04,05,06,07,08,09,10,11 vilket löser både
		det estetiska med att sorteringen är konstig och ovanstående bugg.


assets/php/utils.php
<?php
 
	class Colors {
		private $foreground_colors = array();
		private $background_colors = array();
 
		public function __construct() {
			// Set up shell colors
			$this->foreground_colors['black'] = '0;30';
			$this->foreground_colors['dark_gray'] = '1;30';
			$this->foreground_colors['blue'] = '0;34';
			$this->foreground_colors['light_blue'] = '1;34';
			$this->foreground_colors['green'] = '0;32';
			$this->foreground_colors['light_green'] = '1;32';
			$this->foreground_colors['cyan'] = '0;36';
			$this->foreground_colors['light_cyan'] = '1;36';
			$this->foreground_colors['red'] = '0;31';
			$this->foreground_colors['light_red'] = '1;31';
			$this->foreground_colors['purple'] = '0;35';
			$this->foreground_colors['light_purple'] = '1;35';
			$this->foreground_colors['brown'] = '0;33';
			$this->foreground_colors['yellow'] = '1;33';
			$this->foreground_colors['light_gray'] = '0;37';
			$this->foreground_colors['white'] = '1;37';
 
			$this->background_colors['black'] = '40';
			$this->background_colors['red'] = '41';
			$this->background_colors['green'] = '42';
			$this->background_colors['yellow'] = '43';
			$this->background_colors['blue'] = '44';
			$this->background_colors['magenta'] = '45';
			$this->background_colors['cyan'] = '46';
			$this->background_colors['light_gray'] = '47';
		}
 
		// Returns colored string
		public function getColoredString($string, $foreground_color = null, $background_color = null) {
			$colored_string = "";
 
			// Check if given foreground color found
			if (isset($this->foreground_colors[$foreground_color])) {
				$colored_string .= "\033[" . $this->foreground_colors[$foreground_color] . "m";
			}
			// Check if given background color found
			if (isset($this->background_colors[$background_color])) {
				$colored_string .= "\033[" . $this->background_colors[$background_color] . "m";
			}
 
			// Add string and end coloring
			$colored_string .=  $string . "\033[0m";
 
			return $colored_string;
		}
 
		// Returns all foreground color names
		public function getForegroundColors() {
			return array_keys($this->foreground_colors);
		}
 
		// Returns all background color names
		public function getBackgroundColors() {
			return array_keys($this->background_colors);
		}
	}

    /**
     *  Outputs a $message to STDOUT and log file, color-coded depending on $case.
     *  @param $message - the message to log
     *  @param $case - the type of severity and reson to log, the message will be color-coded accordingly, the cases are:
     *          userConnected, userDisconnected, addedItem, updatedItem, removedItem, clearedList, synced, error, warn
     */
    function logMsg($message, $level = "default") {
        $colors = new Colors();
        $fg = null;
        $bg = null;
        
        switch ($level) {
            case 'userConnected' :
                $fg = null;
                $bg = "cyan";
                break;
            case 'userDisconnected' :
                $fg = "cyan";
                $bg = "black";
                break;
            case 'addedItem' :
                $fg = "black";
                $bg = "green";
                break;
            case 'updatedItem' :
                $fg = "light_gray";
                $bg = "green";
                break;
            case 'clearedList' :
                $fg = "light_gray";
                $bg = "black";
                break;
            case 'removedItem' :
                $fg = "yellow";
                $bg = "black";
                break;
            case 'error' :
                $fg = "white";
                $bg = "red";
                break;
            case 'warn' :
                $fg = "brown";
                $bg = "magenta";
                break;
            case 'sync' :
                $fg = "light_green";
                $bg = "blue";
                break;
            default : 
                break;
        }
        
        echo  $colors->getColoredString($message, $fg, $bg) . "\n<br>";
        echo GetCallingMethodName();
        date_default_timezone_set('Europe/Stockholm');
		file_put_contents("/var/www/shoppinglist/logs/logFile", date("{Y-m-d H:i:s}") . " " . $message . "\n", FILE_APPEND);
    }
    function GetCallingMethodName(){
        $e = new Exception();
        $trace = $e->getTrace();
        //position 0 would be the line that called this function so we ignore it
        $last_call = $trace[1];
        print_r($last_call);
    }
?>